### PR TITLE
Tdt reader

### PR DESCRIPTION
**Pull request by kajal5888**
Author Guidelines
-----------------
- [ ] Is the change set **< 400 lines**?
- [ ] Was the code checked for memory leaks/performance bottlenecks?
- [X] Is the code running locally **and** on the ESI cluster?
- [X] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [x] Are testing routines present?
- [x] Do **parallel** loops have a set length and correct termination conditions?
- [x] Do objects in the global package namespace perform proper parsing of their input? 
- [x] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [x] Does the code adhere to agreed-upon naming conventions?
  - [x] Are keywords clearly named and easy to understand?
  - [x] No commented-out code?
- [x] Are all docstrings complete and accurate?

### CODE REVIEW

**Comment by tensionhead** [comment]
Hey @kajal5888, I adjusted the test script to find the test data again..

File: syncopy/io/load_tdt.py [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+    tdtPaths = []
+    for fname in f_names:
+        fname = os.path.join(directory, fname)
+        f_path, f_name = io_parser(fname, varname="tdt source file name", isfile=True, exists=True)
+        tdtPaths.append(os.path.join(f_path, f_name))
+    tdtPaths = _natural_sort(tdtPaths)
+    return tdtPaths
+
+
+def _natural_sort(file_names):
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:688]: That's a very neat function and solution 👍

File: syncopy/io/load_tdt.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+        self.STROFF = int('00000102', 16)
+        self.SCALAR = int('00000201', 16)
+        self.STREAM = int('00008101', 16)
+        self.SNIP = int('00008201', 16)
+        self.MARK = int('00008801', 16)
+        self.HASDATA = int('00008000', 16)
+        self.UCF = int('00000010', 16)
+        self.PHANTOM = int('00000020', 16)
+        self.MASK = int('0000FF0F', 16)
+        self.INVALID_MASK = int('FFFF0000', 16)
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:64]: this looks like magic 🙂

File: syncopy/io/load_tdt.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+        self.STARTBLOCK = int('0001', 16)
+        self.STOPBLOCK = int('0002', 16)
+        self.ALLOWED_FORMATS = [np.float32, np.int32, np.int16, np.int8, np.float64, np.int64]
+        self.ALLOWED_EVTYPES = ['all', 'epocs', 'snips', 'streams', 'scalars']
+
+    def code_to_type(self, code):
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:70]: strictly this does a `code -> str` mapping no?!

File: syncopy/io/load_tdt.py [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+        else:
+            sample = np.ceil(sample) if t1 else np.round(sample)
+        sample = np.uint64(sample)
+        if to_time:
+            return np.float64(sample) / fs
+        return sample
+
+    def check_ucf(self, code):
+        # given event code, check if it has unique channel files
+        return code & self.UCF == self.UCF
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:105]: what exactly happens here? You bitwise compare two integers??

File: syncopy/io/load_tdt.py [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+        strobe_on_types = [self.STRON, self.MARK]
+        strobe_off_types = [self.STROFF]
+        if code in strobe_on_types:
+            return 'onset'
+        elif code in strobe_off_types:
+            return 'offset'
+        return 'unknown'
+
+    def code_to_name(self, code):
+        return int(code).to_bytes(4, byteorder = 'little').decode('cp437')
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:118]: wow.. for such a line a little one liner comment what is about to happen always helps fellow programmers ;)

File: syncopy/io/load_tdt.py [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+
+    def get_files(self, ext, speci_trgt):
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:120]: such a general helper routine (also useful for the other class) can also be moved outside the class to `_get_files`

File: syncopy/io/load_tdt.py [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+        if header.stop_time > 0:
+            data.info.duration = data.info.stop_date - data.info.start_date  # datestr(s2-s1, 'HH:MM:SS')
+
+        tsq.seek(40, os.SEEK_SET)
+
+        read_size = 10000000 if self.t2 > 0 else 50000000
+        header.stores = StructDict()
+        code_ct = 0
+        while True:
+            heads = np.frombuffer(tsq.read(read_size * 4), dtype = np.uint32)
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:196]: wow.. this looks like C 👍

File: syncopy/io/load_tdt.py [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+                    'ucf': self.check_ucf(heads[1, unique_ind[counter]]),
+                    'epoc_type': self.epoc_to_type(heads[1, unique_ind[counter]]),
+                    'dform': heads[8, unique_ind[counter]],
+                    'size': heads[0, unique_ind[counter]],
+                    'buddy': heads[3, unique_ind[counter]],
+                    'temp': heads[:, unique_ind[counter]]
+                })
+
+            # Looking for only Mark, PDi\ and PDio
+            looking_for = ["Mark", "PDio", 'LFPs', "PDi\\"]  #
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:232]: Could this be good to expose to the users? If they want to look for other targets?

File: syncopy/io/load_tdt.py [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+
+            for tar in targets.items():
+                store_code = store_codes[tar[1]]
+                store_code['name'] = self.code_to_name(store_code['code'])
+                # print(code_to_name(store_code['code']), store_code['code'])
+
+                store_code['var_name'] = store_code['name']
+                var_name = store_code['var_name']
+
+                if store_code['type_str'] == 'epocs':
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:246]: there is no else, what exactly is done here?

File: syncopy/io/load_tdt.py [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+                    vvv = self.time2sample(vvv, to_time = True)
+                    header.stores[var_name].ts.append(vvv)
+                    if (not self.nodata) or (store_code['type_str'] == 'streams'):
+                        if not hasattr(header.stores[var_name], 'data'):
+                            header.stores[var_name].data = []
+                        header.stores[var_name].data.append(np.reshape(heads[[[6], [7]], valid_ind].T, (-1, 1)).T.view(np.float64))
+                    if not hasattr(header.stores[var_name], 'chan'):
+                        header.stores[var_name].chan = []
+                    header.stores[var_name].chan.append(temp[::2])
+                else:
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:300]: ah here is an `else`.. mmh, I guess the logical program flow could be optimized here

File: syncopy/io/load_tdt.py [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+                    loc = epocs.name.index(store_code['name'])
+                    # round timestamps to the nearest sample
+                    vvv = np.reshape(heads[[[4], [5]], valid_ind].T, (-1, 1)).T.view(np.float64) - header.start_time
+                    # round timestamps to the nearest sample
+                    vvv = self.time2sample(vvv, to_time = True)
+                    epocs.ts[loc] = np.append(epocs.ts[loc], vvv)
+                    epocs.data[loc] = np.append(epocs.data[loc], np.reshape(heads[[[6], [7]], valid_ind].T, (-1, 1)).T.view(np.float64))
+            last_ts = heads[[4, 5], -1].view(np.float64) - header.start_time
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:308]: Also here, without comments it's hard to guess what happens

File: syncopy/io/load_tdt.py [hunk 12]
```diff
@@ -0,0 +1,700 @@
...
+            if 'sortcode' in header.stores[var_name].keys():
+                header.stores[var_name].sortcode = np.concatenate(header.stores[var_name].sortcode)
+            if 'data' in header.stores[var_name].keys():
+                if header.stores[var_name].type_str != 'epocs':
+                    header.stores[var_name].data = np.concatenate(header.stores[var_name].data, axis = 1)[0]
+
+            # if it's a data type, cast as a file offset pointer instead of data
+            if header.stores[var_name].type_str in ['streams', 'snips']:
+                if 'data' in header.stores[var_name].keys():
+                    header.stores[var_name].data = header.stores[var_name].data.view(np.uint64)
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:370]: uff.. also hard to read, what is happening here?

File: syncopy/io/load_tdt.py [hunk 13]
```diff
@@ -0,0 +1,700 @@
...
+            if 'chan' in header.stores[var_name].keys():
+                if np.max(header.stores[var_name].chan) == 1:
+                    header.stores[var_name].chan = [1]
+
+        valid_time_range = np.array([[self.t1], [self.t2]]) if self.t2 > 0 else np.array([[self.t1], [np.inf]])
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:375]: This is nicely concise yet still readable imho 👍

File: syncopy/io/load_tdt.py [hunk 14]
```diff
@@ -0,0 +1,700 @@
...
+        ranges = None
+        if hasattr(ranges, "__len__"):
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:377]: how is this gonna work? You set `ranges = None` right before 😅

File: syncopy/io/load_tdt.py [hunk 15]
```diff
@@ -0,0 +1,700 @@
...
+                                ind2 = ind2[-1]
+                                # keep one prior for streams (for all channels)
+                                nchan = max(data[current_type_str][var_name].chan)
+                                if ind2 - nchan >= -1:
+                                    filter_ind[jj] = ind2 - np.arange(nchan - 1, -1, -1)
+                                    temp = data[current_type_str][var_name].ts[filter_ind[jj]]
+                                    data[current_type_str][var_name].start_time[jj] = temp[0]
+                                    bSkip = 1
+
+                    if len(filter_ind[jj]) > 0:
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:426]: what exactly get achieved in the following ~20 lines inside this `if` block?

File: syncopy/io/load_tdt.py [hunk 16]
```diff
@@ -0,0 +1,700 @@
...
+            sz = np.uint64(np.dtype(current_data_format).itemsize)
+
+            if current_type_str == 'scalars':
+                if len(data[current_type_str][current_name].chan) > 0:
+                    nchan = int(np.max(data[current_type_str][current_name].chan))
+                else:
+                    nchan = 0
+                if nchan > 1:
+                    # organize data by sample
+                    # find channels with most and least amount of data
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:534]: why is this necessary? What happens with that information?

File: syncopy/io/load_tdt.py [hunk 17]
```diff
@@ -0,0 +1,700 @@
...
+            data = np.fromfile(f, dtype = 'single')
+        return data
+
+    def md5sum(self, filename):
+        from hashlib import md5
+        hash = md5()
+        with open(filename, "rb") as f:
+            for chunk in iter(lambda: f.read(128 * hash.block_size), b""):
+                hash.update(chunk)
+        return hash.hexdigest()
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:598]: Nice and clean 👍

File: syncopy/io/load_tdt.py [hunk 18]
```diff
@@ -0,0 +1,700 @@
...
+
+    def data_aranging(self, Files, DataInfo_loaded):
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:600]: naming is hard.. 😛

File: syncopy/io/load_tdt.py [hunk 19]
```diff
@@ -0,0 +1,700 @@
...
+            idxStartStop = [np.clip(np.array((jj, jj + self.chan_in_chunks)),
+                                    a_min = None, a_max = len(Files))
+                            for jj in range(0, len(Files), self.chan_in_chunks)]
+            print("Merging {0} files in {1} chunks each with {2} channels into \n   {3}".format(
+                len(Files), len(idxStartStop), self.chan_in_chunks,
+                hdf_out_path))
+            for (start, stop) in tqdm(iterable = idxStartStop, desc = "chunk", unit = "chunk"):
+                data = [self.read_data(Files[jj]) for jj in range(start, stop)]
+                data = np.vstack(data).T
+                if start == 0:
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:617]: ok so the initial read sets the shape for all following reads, nicely solved 👍

File: syncopy/io/load_tdt.py [hunk 20]
```diff
@@ -0,0 +1,700 @@
...
+            # link dataset to AnalogData instance
+            AData.data = target# [:, chanind]
+        
+        chanlist = None if self.channels == 'all' else ['channel'+str(trch+1).zfill(3) for trch in self.channels]
+        chanind = np.arange(len(Files)) if self.channels == 'all' else self.channels        
+        AData.samplerate = DataInfo_loaded.LFPs.fs
+        AData.channel = chanlist
+
+        # write info file
+        AData.cfg["originalFiles"] = Files,
```
Comment (tensionhead) [inline syncopy/io/load_tdt.py:636]: here I will replace all `cfg` with our new `.info` property 🙂

**Review by tensionhead — COMMENTED** [review]
Thanks a lot @kajal5888 ! This looks quite intricate, see my detailed comments where commenting on the code might be beneficial for future readers. I can't say I fully unerstand what is going on right now, so if I can ping you as our **Syncopy TDT reader maintainer** in the future, would be great 🙂 
As discussed (and partly already done) I will wrap your two classes into one user exposed `load_tdt` function and also add some tests to `test_spyio.py`. Then it should be ready to go!

**Comment by tensionhead** [comment]
Ok I think it's ready for a merge, @dfsp-spirit could you have a final look over all this? Thx!

File: syncopy/io/load_tdt.py [hunk 21]
```diff
@@ -0,0 +1,860 @@
...
+def load_tdt(data_path, start_code=None, end_code=None,
+             subtract_median=False):
+
+    """
+    Imports TDT time series data and meta-information
+    into a single :class:`~syncopy.AnalogData` object.
+
+    An ad-hoc trialdefinition will be attached if
+    both `start_code` and `end_code` are given.
+    Otherwise a single all-to-all trialdefintion
```
Comment (dfsp-spirit) [inline syncopy/io/load_tdt.py:34]: typo in "trialdefintion"

File: syncopy/io/load_tdt.py [hunk 22]
```diff
@@ -0,0 +1,860 @@
...
+    >>> trg_sample = adata.info['Trigger_sample']
+
+    Load the same data and construct a trialdefinition on the fly:
+
+    >>> adata = load_tdt('/data/session3', start_code=23000, end_code=30020)
+
+    Access the 3rd trial:
+
+    >>> trl_dat = adata.trials[2]
+    """
```
Comment (dfsp-spirit) [inline syncopy/io/load_tdt.py:82]: nice that we have examples!

File: syncopy/io/load_tdt.py [hunk 23]
```diff
@@ -0,0 +1,860 @@
...
+    io_parser(data_path, isfile=False)
+
+    if start_code is not None and end_code is None:
+        lgl = "trigger codes for both trial start and end"
+        raise SPYValueError(lgl, "end_code", end_code)
+    if end_code is not None and start_code is None:
+        lgl = "trigger codes for both trial start and end"
+        raise SPYValueError(lgl, "start_code", start_code)
+    elif end_code is None and start_code is None:
+        pass
```
Comment (dfsp-spirit) [inline syncopy/io/load_tdt.py:93]: superfluous I guess, unless you want it here do document the interface. then I would not use `if..., if..elif`but `if, elif, else`.

**Comment by dfsp-spirit** [comment]
I don't know the format in detail and I guess testing with real-world data is important here.

Ready to merge after addressing minor comments.
